### ISSUE TITLE
App hangs after loading a project

### ISSUE THREAD

**Issue by cckozie** [unfoldingWord/translationCore#1307]
Imported https://door43.org/u/qa98/ar_mat_text_ulb
Opened the project, selected tW. The project opened but the app is then frozen. Could not open the console or do anything else. Had to close the app. Repeatable.

**Comment by mannycolon** [comment]
@klappy this bug seems to be related to the fact that we are fetching the data in componentwill mount() 
we are not encountering this bug as often because we are using smaller book/projects for testing but once we use a bigger project this issue comes up.

**Comment by benjore** [comment]
- [ ] SPIKE

**Comment by EllDoubleYew** [comment]
So some important nuances here. Does the app ever load? Or is it truly completely frozen?

I tested this and my app got stuck with the loading screen up before eventually getting a white screen and my dev tools disconnecting. I am not sure if this is consistent or not.

This is a very difficult problem to analyze because we have essentially no hints on why this is happening. I feel like this is where some redux analysis tools might be handy to get a deeper understanding of what is causing this.

**Comment by EllDoubleYew** [comment]
I also want to not here the problem could be trivial. The problem could be gargantuan. However we need to find a way to find the problem. There are really multiple layers of research involved here. We need to find a tool that allows us to analyze the app when it is in this "locked up" state. And THEN we need to use that to find the issue with the app. Two different things we need to find before we can begin addressing the problem with any sort of organized approach.

**Comment by EllDoubleYew** [comment]
![image](https://cloud.githubusercontent.com/assets/7133050/25720730/e7330f4a-30db-11e7-9a31-f6273cbec29d.png)

Some quick findings. There is a gigantic misuse of system memory here or a gigantic memory leak. This is what The Chrome profiler found while running an allocation timeline over the course of loading a Psalms projects. This is 6 GB of RAM it is using somewhere to store "System Objects".

The Heap Timeline took about 20 minutes to gather and I have it saved (its a 6 GB file). I have no clue how to analyze this in any meaningful way and am quickly stepping out of my depth. Maybe @klappy has some experience with this?

**Comment by klappy** [comment]
My gut is telling me that we need to explore breaking down resources, like the bible into smaller files and store them in book/chapter/verse folder structure like we do for checkData. Then we load the verses into memory as needed just like we do for comments/etc... 

It doesn't sound like a lot when we look at a large json file as it looks like it fits into memory okay. But with the little bit that I do with the Bible in TACT/MAP, it consumes 2-4GB of memory just aligning the bible's text even though internally it is just text, the in memory representation of that data is so much heavier. I'm not even doing anything as fancy as we are with the Scripture Pane in tC.

**Comment by klappy** [comment]
@EllDoubleYew @mannycolon: I found the root issue for the project that Koz brought up as being hung. tS projects have more than just chunk/verse files in the chapter folders. When I delete the files that hold the titles/front/back.json, then the project loads without issue. Somehow tC is stuck in a loop when trying to write ./apps/translationCore/resources/bibles/targetLanguage.json. That's when I started debugging and deleting files in the project that might cause an issue. 

We will need to filter and only read in files that have a digit, so use a regex or something to filter the right files when reading the tS project into the targetLanguage bible.

**Comment by klappy** [comment]
There appears to be another issue to where the first loading of that project is good and renders now. The second time, it gets stuck again. While it's stuck I explored the filesystem and saw the each bible being overwritten in a loop with a zero byte file. Finder flashes the file info pane when a file is written to so it can recalculate file size. They all flash while it's stuck, so it feels like it is caught in the Bible functions.

**Comment by EllDoubleYew** [comment]
Putting this back on the cooler for a minute because I don't know if I can finish anything that would remedy this by review. 

@klappy Have you seen the "too many files open" error? I am not 100% sure what is going on with that but I think it may be related. My short sighted approach was to always open a project like you were opening it for the first time (aka fetching it instead of actually loading it every time). But a.) this wont save any changes you make to it, at least in any meaningful way and b.) what books/projects would I do this for? How big is too big? What is the tipping point in terms of books?

I have realized how easy we have had it only having to worry about eph and tit right now. Full bible support is hard 🙁 .

Let me know if you want to look at some of this with me one day.

**Comment by cckozie** [comment]
Hopefully will be fixed in #1499

**Comment by mannycolon** [comment]
@cckozie i believe this issue was addressed by #1499

**Comment by cckozie** [comment]
0.6.26.4
The project now does not load. It stays on the loading dialog for a long time, then eventually the entire tC window goes blank. During the loading I noticed that the tC process was taking 1.7GB+ of memory on a 4GB machine.
Nothing is logged in the console during the loading process.
I have not seen this type of failure on Ephesians or Titus projects.

**Comment by mannycolon** [comment]
@cckozie what project did you use?

**Comment by mannycolon** [comment]
Make sure to delete the apps folder from the project folder

<img width="1199" alt="screen shot 2017-05-19 at 1 42 19 pm" src="https://cloud.githubusercontent.com/assets/17072300/26259892/14073474-3c99-11e7-944d-7888f5183027.png">

**Comment by mannycolon** [comment]
on my end it loads but it does take a long time.

**Comment by cckozie** [comment]
I did delete the .apps folder (it didn't have one after the import) and tried it again. It displays the loading screen for about 90 seconds, then crashes with an empty window.
[ar_mat_text_ulb.zip](https://github.com/unfoldingWord-dev/translationCore/files/1015297/ar_mat_text_ulb.zip)

**Comment by mannycolon** [comment]
@cckozie I believe I made changes that fixed this problem because I can't recreate it. This changes were not included in 0.6.26.4 so I'm hoping those changes that I made addressed this issue.

**Comment by cckozie** [comment]
@mannycolon OK, I'll put a QA Hold on it and check with the next build.

**Comment by cckozie** [comment]
Or never mind.

**Comment by cckozie** [comment]
The project does open in 0.26.4.5, but it takes more than 3 minutes in my Mac. Passing this since we're officially supporting only Ephesians and Titus now, and writing up a new issue. #1655
